Implement multi-stage build process for the Docker image

vbrandl opened the pull request:
Multi-stage builds help producing smaller image sizes. This setup uses
the `ekidd/rust-musl-builder` image to compile the source code and
executes the benchmark in an Alpine Linux image.

The old `Dockerfile` produced an `1.81GB` image while the resulting
image using the new `Dockerfile` is only `10.1MB`.

(As discussed in #14)

vbrandl committed:
Implement multi-stage build process for the Docker image

Multi-stage builds help producing smaller image sizes. This setup uses
the `ekidd/rust-musl-builder` image to compile the source code and
executes the benchmark in an Alpine Linux image.

The old `Dockerfile` produced an `1.81GB` image while the resulting
image using the new `Dockerfile` is only `10.1MB`.

vbrandl committed:
Merge branch 'master' into feature/docker-multi-stage-build

mre commented:
Quite an improvement!

Tested it on my machine like so:

```
> docker system prune -a           
> docker build -t fcat .
Sending build context to Docker daemon  149.7MB
Step 1/12 : FROM ekidd/rust-musl-builder:1.27.2 as builder
1.27.2: Pulling from ekidd/rust-musl-builder
8ee29e426c26: Pull complete
6e83b260b73b: Pull complete
e26b65fd1143: Pull complete
40dca07f8222: Pull complete
b420ae9e10b3: Pull complete
3a28293b993c: Pull complete
e2d903666f97: Pull complete
fb412d0f8c5e: Pull complete
0e6d3107c55b: Pull complete
81610f7c1cac: Pull complete
27645b03e728: Pull complete
c020c02f573c: Pull complete
1108b96d2ad7: Pull complete
3f904018804a: Pull complete
Digest: sha256:dce1c2f06a39e2b686b0fe415b3f2778659e27a6f2f09ecdfc5630ddc3e6b772
Status: Downloaded newer image for ekidd/rust-musl-builder:1.27.2
 ---> 9c20e21a37dc
Step 2/12 : RUN USER=rust cargo new --bin fcat
 ---> Running in 6c9ca00c6f0b
     Created binary (application) `fcat` project
Removing intermediate container 6c9ca00c6f0b
 ---> 20ab114eb840
Step 3/12 : WORKDIR /home/rust/src/fcat
 ---> Running in 5c2d979f771c
Removing intermediate container 5c2d979f771c
 ---> c15ee02452dd
Step 4/12 : RUN touch -t 197001010000 src/main.rs
 ---> Running in b238780294f2
Removing intermediate container b238780294f2
 ---> 25f2e3764638
Step 5/12 : COPY ./Cargo.lock Cargo.lock
 ---> 4f2958c9d870
Step 6/12 : COPY ./Cargo.toml Cargo.toml
 ---> cd56008cc71a
Step 7/12 : RUN cargo build --release
 ---> Running in c9fce863f0f7
 Downloading nix v0.9.0
 Downloading cfg-if v0.1.2
 Downloading bitflags v0.9.1
 Downloading libc v0.2.36
 Downloading void v1.0.2
   Compiling void v1.0.2
   Compiling libc v0.2.36
   Compiling cfg-if v0.1.2
   Compiling bitflags v0.9.1
   Compiling nix v0.9.0
   Compiling fcat v0.1.0 (file:///home/rust/src/fcat)
    Finished release [optimized] target(s) in 8.54s
Removing intermediate container c9fce863f0f7
 ---> 0ee0d4eff4ad
Step 8/12 : FROM alpine:latest
latest: Pulling from library/alpine
8e3ba11ec2a2: Pull complete
Digest: sha256:7043076348bf5040220df6ad703798fd8593a0918d06d3ce30c6c93be117e430
Status: Downloaded newer image for alpine:latest
 ---> 11cd0b38bc3c
Step 9/12 : RUN apk --update --no-cache add pv && rm -rf /var/cache/apk/*
 ---> Running in 24d3609e9203
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/community/x86_64/APKINDEX.tar.gz
(1/1) Installing pv (1.6.6-r1)
Executing busybox-1.28.4-r0.trigger
OK: 5 MiB in 14 packages
Removing intermediate container 24d3609e9203
 ---> b1cf8d1cc614
Step 10/12 : COPY --from=builder /home/rust/src/fcat/target/x86_64-unknown-linux-musl/release/fcat /usr/local/sbin/fcat
 ---> 2d02a5cd259f
Step 11/12 : COPY entrypoint /entrypoint
 ---> 4fe201abccd1
Step 12/12 : ENTRYPOINT ["/entrypoint"]
 ---> Running in 36d5c25bf323
Removing intermediate container 36d5c25bf323
 ---> aba5d26a01d4
Successfully built aba5d26a01d4
Successfully tagged fcat:latest
> docker run  -it --entrypoint sh fcat
# /usr/local/sbin/fcat
Hello, world!
```

vbrandl committed:
Fix caching problems

mre commented on Dockerfile:
```diff
@@ -3,15 +3,19 @@ FROM ekidd/rust-musl-builder:1.27.2 as builder
 
 # this build step will cache your dependencies
 RUN cargo build --release
+# remove cached build artefact to prevent caching issues
```
artefact -> artifact

vbrandl replied:
Fixed in b297679

mre commented:
Works now. Thanks! Minor nitpick in the comments before I'd like to merge this. Good work!

vbrandl committed:
Fix typo in comments

vbrandl committed:
Use `sh` as shell interpreter

Alpine Linux doesn't come with `bash` pre-installed but the `entrypoint`
script executes just fine using `sh` as interpreter

vbrandl commented:
At least inside the Docker container, fcat is not that much faster than regular cat:
```
$ docker run  -it --rm local/fcat                                                                                                                          
mount: permission denied (are you root?)
Creating a kitty large benchmark file
6000+0 records in
6000+0 records out
Benchmarking system cat:
[ 169MiB/s]
Benchmarking fcat:
[ 204MiB/s]
```

mre commented:
Hum... I wonder why. About the same on my machine. We should investigate maybe, but maybe in a separate issue.
I guess we can merge your changes, right? Looks good from my side.

vbrandl commented:
Yeah I think you can merge now
